diff --git a/code/AssetLib/Obj/ObjExporter.cpp b/code/AssetLib/Obj/ObjExporter.cpp
index 1622bc4..e4db29d 100644
--- a/code/AssetLib/Obj/ObjExporter.cpp
+++ b/code/AssetLib/Obj/ObjExporter.cpp
@@ -178,6 +178,9 @@ std::string ObjExporter::GetMaterialName(unsigned int index) {
     if ( nullptr == pScene->mMaterials ) {
         return EmptyStr;
     }
+    if (index >= pScene->mNumMaterials) {
+        return EmptyStr;
+    }
     const aiMaterial* const mat = pScene->mMaterials[index];
     if ( nullptr == mat ) {
         return EmptyStr;
@@ -307,7 +310,7 @@ void ObjExporter::WriteGeometryFile(bool noMtl, bool merge_identical_vertices) {
         if (!m.name.empty()) {
             mOutput << "g " << m.name << endl;
         }
-        if ( !noMtl ) {
+        if ( !noMtl && !m.matname.empty() ) {
             mOutput << "usemtl " << m.matname << endl;
         }
 
diff --git a/code/AssetLib/Obj/ObjFileData.h b/code/AssetLib/Obj/ObjFileData.h
index 0d43678..87863b9 100644
--- a/code/AssetLib/Obj/ObjFileData.h
+++ b/code/AssetLib/Obj/ObjFileData.h
@@ -168,6 +168,8 @@ struct Material {
     aiColor3D diffuse;
     //! Specular color
     aiColor3D specular;
+    //! True when the MTL explicitly defines an emissive color.
+    bool hasEmissive = false;
     //! Emissive color
     aiColor3D emissive;
     //! Alpha value
diff --git a/code/AssetLib/Obj/ObjFileImporter.cpp b/code/AssetLib/Obj/ObjFileImporter.cpp
index f169922..4f7feb7 100644
--- a/code/AssetLib/Obj/ObjFileImporter.cpp
+++ b/code/AssetLib/Obj/ObjFileImporter.cpp
@@ -621,7 +621,9 @@ void ObjFileImporter::createMaterials(const ObjFile::Model *pModel, aiScene *pSc
         mat->AddProperty(&pCurrentMaterial->ambient, 1, AI_MATKEY_COLOR_AMBIENT);
         mat->AddProperty(&pCurrentMaterial->diffuse, 1, AI_MATKEY_COLOR_DIFFUSE);
         mat->AddProperty(&pCurrentMaterial->specular, 1, AI_MATKEY_COLOR_SPECULAR);
-        mat->AddProperty(&pCurrentMaterial->emissive, 1, AI_MATKEY_COLOR_EMISSIVE);
+        if (pCurrentMaterial->hasEmissive) {
+            mat->AddProperty(&pCurrentMaterial->emissive, 1, AI_MATKEY_COLOR_EMISSIVE);
+        }
         mat->AddProperty(&pCurrentMaterial->shineness, 1, AI_MATKEY_SHININESS);
         mat->AddProperty(&pCurrentMaterial->alpha, 1, AI_MATKEY_OPACITY);
         mat->AddProperty(&pCurrentMaterial->transparent, 1, AI_MATKEY_COLOR_TRANSPARENT);
diff --git a/code/AssetLib/Obj/ObjFileMtlImporter.cpp b/code/AssetLib/Obj/ObjFileMtlImporter.cpp
index 02ccfcb..96b199b 100644
--- a/code/AssetLib/Obj/ObjFileMtlImporter.cpp
+++ b/code/AssetLib/Obj/ObjFileMtlImporter.cpp
@@ -149,8 +149,10 @@ void ObjFileMtlImporter::load() {
                         getColorRGBA(&m_pModel->mCurrentMaterial->specular);
                 } else if (*m_DataIt == 'e') {
                     ++m_DataIt;
-                    if (m_pModel->mCurrentMaterial != nullptr)
+                    if (m_pModel->mCurrentMaterial != nullptr) {
+                        m_pModel->mCurrentMaterial->hasEmissive = true;
                         getColorRGBA(&m_pModel->mCurrentMaterial->emissive);
+                    }
                 }
                 m_DataIt = skipLine<DataArrayIt>(m_DataIt, m_DataItEnd, m_uiLine);
             } break;
diff --git a/code/AssetLib/Obj/ObjFileParser.cpp b/code/AssetLib/Obj/ObjFileParser.cpp
index f1f0b40..297fc8f 100644
--- a/code/AssetLib/Obj/ObjFileParser.cpp
+++ b/code/AssetLib/Obj/ObjFileParser.cpp
@@ -303,7 +303,8 @@ size_t ObjFileParser::getNumComponentsInDataDefinition() {
         if (!SkipSpaces(&tmp, mEnd) || *tmp == '#') {
             break;
         }
-        const bool isNum(IsNumeric(*tmp) || isNanOrInf(tmp));
+        const bool isNum(IsNumeric(*tmp) || isNanOrInf(tmp) ||
+                (*tmp == '.' && tmp + 1 < mEnd && !IsLineEnd(*(tmp + 1)) && IsNumeric(*(tmp + 1))));
         SkipToken(tmp, mEnd);
         if (isNum) {
             ++numComponents;
diff --git a/code/PostProcessing/CalcTangentsProcess.cpp b/code/PostProcessing/CalcTangentsProcess.cpp
index 3149079..2b8ada4 100644
--- a/code/PostProcessing/CalcTangentsProcess.cpp
+++ b/code/PostProcessing/CalcTangentsProcess.cpp
@@ -286,8 +286,8 @@ bool CalcTangentsProcess::ProcessMesh(aiMesh *pMesh, unsigned int meshIndex) {
             smoothTangent += meshTang[closeVertices[b]];
             smoothBitangent += meshBitang[closeVertices[b]];
         }
-        smoothTangent.Normalize();
-        smoothBitangent.Normalize();
+        smoothTangent.NormalizeSafe();
+        smoothBitangent.NormalizeSafe();
 
         // and write it back into all affected tangents
         for (unsigned int b = 0; b < closeVertices.size(); ++b) {
diff --git a/include/assimp/IOStreamBuffer.h b/include/assimp/IOStreamBuffer.h
index 29b7cf3..6e36a64 100644
--- a/include/assimp/IOStreamBuffer.h
+++ b/include/assimp/IOStreamBuffer.h
@@ -234,45 +234,57 @@ AI_FORCE_INLINE size_t IOStreamBuffer<T>::getFilePos() const {
 
 template <class T>
 AI_FORCE_INLINE bool IOStreamBuffer<T>::getNextDataLine(std::vector<T> &buffer, T continuationToken) {
-    buffer.resize(m_cacheSize);
-    if (m_cachePos >= m_cacheSize || 0 == m_filePos) {
-        if (!readNextBlock()) {
-            return false;
-        }
-    }
+    buffer.clear();
+    bool inContinuationState = false;
+    while (true) {
+        bool hasLineEndings = false;
+        while (m_cachePos >= m_cacheSize || 0 == m_filePos) {
+            if (!readNextBlock()) {
+                break;
+            }
 
-    size_t i = 0;
-    for (;;) {
-        if (continuationToken == m_cache[m_cachePos] && IsLineEnd(m_cache[m_cachePos + 1])) {
-            ++m_cachePos;
-            while (m_cache[m_cachePos] != '\n') {
-                ++m_cachePos;
+            // Skip empty lines.
+            while (m_cachePos < m_cacheSize && IsLineEnd(m_cache[m_cachePos])) {
+                hasLineEndings = true;
+                m_cachePos++;
             }
             ++m_cachePos;
-        } else if (IsLineEnd(m_cache[m_cachePos])) {
-            break;
         }
-
-        buffer[i] = m_cache[m_cachePos];
-        ++m_cachePos;
-        ++i;
-
-        if(i == buffer.size()) {
-            buffer.resize(buffer.size() * 2);
-        }
+        if (m_cachePos >= m_cacheSize) {
+            break;
+        }
 
-        if (m_cachePos >= size()) {
+        // End this line if we hit line endings outside a continuation state.
+        if (hasLineEndings && !inContinuationState && buffer.size() > 0) {
             break;
         }
+
         if (m_cachePos >= m_cacheSize) {
-            if (!readNextBlock()) {
-                return false;
+            break;
+        } else if (inContinuationState) {
+            // Skip all line endings while inside a continuation.
+            if (IsLineEnd(m_cache[m_cachePos])) {
+                ++m_cachePos;
+            } else {
+                inContinuationState = false;
+            }
+        } else if (!IsLineEnd(m_cache[m_cachePos])) {
+            buffer.push_back(m_cache[m_cachePos]);
+            ++m_cachePos;
+        } else {
+            ++m_cachePos;
+            if (buffer.size() != 0) {
+                // Continue reading the next line if this one ended with the continuation token.
+                if (buffer.back() == continuationToken) {
+                    inContinuationState = true;
+                    buffer.pop_back();
+                } else {
+                    break;
+                }
             }
         }
     }
+    buffer.push_back('\n');
 
-    buffer[i] = '\n';
-    ++m_cachePos;
-
-    return true;
+    return buffer.size() > 1;
 }
